tests/cpp_tests/test_stream.cpp

```diff
@@ -0,0 +1,337 @@
+  DatasetHandle dataset_handle = nullptr;
+  Dataset* dataset = nullptr;
+
+  int has_weights = weights != nullptr;
+  int has_init_scores = init_scores != nullptr;
+  int has_queries = groups != nullptr;
+
+  try {
+    int result = 0;
+    switch (creation_type) {
+      case 0: {
+        Log::Info("Creating Dataset using LGBM_DatasetCreateFromSampledColumn, %d rows dense data with a batch size of %d", nrows, batch_count);
+
+        // construct sample data first (use all data for convenience and since size is small)
+        std::vector<std::vector<double>> sample_values(ncols);
+        std::vector<std::vector<int>> sample_idx(ncols);
+        const double* current_val = features->data();
+        for (size_t idx = 0; idx < nrows; ++idx) {
+          for (size_t k = 0; k < ncols; ++k) {
+            if (std::fabs(*current_val) > 1e-35f || std::isnan(*current_val)) {
+              sample_values[k].emplace_back(*current_val);
+              sample_idx[k].emplace_back(static_cast<int>(idx));
+            }
+            current_val++;
+          }
+        }
+
+        std::vector<int> sample_sizes;
+        std::vector<double*> sample_values_ptrs;
+        std::vector<int*> sample_idx_ptrs;
+        for (size_t i = 0; i < ncols; ++i) {
+          sample_values_ptrs.push_back(sample_values[i].data());
+          sample_idx_ptrs.push_back(sample_idx[i].data());
+          sample_sizes.push_back(static_cast<int>(sample_values[i].size()));
+        }
+
+        result = LGBM_DatasetCreateFromSampledColumn(
+          sample_values_ptrs.data(),
+          sample_idx_ptrs.data(),
+          ncols,
+          sample_sizes.data(),
+          nrows,
+          nrows,
+          "max_bin=15",
+          &dataset_handle);
+        EXPECT_EQ(0, result) << "LGBM_DatasetCreateFromSampledColumn result code: " << result;
+
+        result = LGBM_DatasetInitStreaming(dataset_handle, has_weights, has_init_scores, has_queries, nclasses, 1);
+        EXPECT_EQ(0, result) << "LGBM_DatasetInitStreaming result code: " << result;
+        break;
+      }
+
+      case 1:
+        Log::Info("Creating Dataset using LGBM_DatasetCreateByReference, %d rows dense data with a batch size of %d", nrows, batch_count);
+        result = LGBM_DatasetCreateByReference(ref_datset_handle, nrows, &dataset_handle);
+        EXPECT_EQ(0, result) << "LGBM_DatasetCreateByReference result code: " << result;
+        break;
+    }
+
+    dataset = static_cast<Dataset*>(dataset_handle);
+
+    Log::Info("Streaming dense dataset, %d rows dense data with a batch size of %d", nrows, batch_count);
+    TestUtils::StreamDenseDataset(
+      dataset_handle,
+      nrows,
+      ncols,
+      nclasses,
+      batch_count,
+      features,
+      labels,
+      weights,
+      init_scores,
+      groups);
+
+    dataset->FinishLoad();
+
+    // TODO(svotaw) we should assert actual feature data, but we would need to calculate bin values
```

Comment:
Please convert this TODO record into a normal GitHub issue.

Reply:
I will just remove this.  I put this TODO early on when thinking about testing thoroughly, but it's mostly referring to testing the existing PushRows functionality for feature data (as opposed to metadata).  It is not actually testing anything new in this PR.  There are actually python tests for that already that I just didn't see at first (and I'm not setup for python anyways so I couldn't use).